### ISSUE TITLE
Ltac's type of is very slow

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#6541]
#### Version

8.8+alpha (e2d1c676b23a335b4fb8a528c99dfca2b82a1a39)

#### Operating system

Linux

#### Description of the problem

(See Also #3280)

Calling `type of` on a closed term in the goal can be 12x slower than calling `set (k := <that term>)`, despite the fact that the latter obviously needs to do some sort of typing.  I realize that `type of` needs to do some extra work to allow it to propagate universe constraints for things like
``coq
Goal True.
  let k := match (eval pattern nat, O in 0) with (fun N : Set => ?f) _ _ => constr:(fun N : Type => f) end in
  let _ := type of k in
  pose k.
```
but does it really need to be 12x slower than `set` on terms with no universes other than `Set` and `Prop`?

```coq
Fixpoint twos (n : nat) (A : Set) : Set :=
  match n with
    | 0 => A
    | S n' => twos n' (prod A A)
  end.

Tactic Notation "display" string(tacname) tactic3(tac) :=
  match goal with |- let n := ?k in _ => idtac k "Time" tacname end;
  lazy [twos];
  time tac ().

Notation "'hide'" := (_ = _) : type_scope.

Goal let n := 21 in twos n nat = twos n nat :> Set.
  display "type of" (fun _ => match goal with |- ?G => let k := type of G in idtac k end). (* 13.197 s *) Undo.
  display "set" (fun _ => match goal with |- ?G => set (G' := G) end). (* 1.127 s *) Undo.
```

(Interestingly, if I stick `Axiom prod : Set -> Set -> Set.` at the top, `type of` is about 2x faster.  But still around 5x slower than `set`.)

**Comment by ppedrot** [comment]
No, `set` doesn't need to retypecheck the term, which is why it's fast on your example. If the original term is well-typed, let-abstraction is always valid (modulo a few gotchas like template polymorphism).

I think I can close this bug as invalid thus.

**Comment by JasonGross** [comment]
Are you saying that `type of` retypechecks the term?  Can there be a variant that doesn't do this?

**Comment by herbelin** [comment]
@JasonGross: yes, Ltac's "type" retypechecks. Coq internally uses another typing function when a term is already known to be well-typed, and this function is as quick as it could (no conversion is done). We could make it available user-side (maybe it is already available from Ltac2 actually).
